File: api/solicitudes.js
import { clientWithJWT, getUserFromRequest } from './_lib/supabase.js'
import { CORS, error, mapSolicitud } from './_lib/helpers.js'

export default async function handler(req, res) {
  if (req.method === 'OPTIONS') {
    return res.setHeader('Access-Control-Allow-Origin', '*')
      .setHeader('Access-Control-Allow-Methods', 'GET,POST,OPTIONS')
      .setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization')
      .status(204).end()
  }
  Object.entries(CORS).forEach(([k, v]) => res.setHeader(k, v))

  const auth = await getUserFromRequest(req)
  if (!auth) return error(res, 'Token requerido', 401)

  const { user, token } = auth
  const sb = clientWithJWT(token)

  const { solicitud_id, action } = req.query

  // POST /api/solicitudes/[solicitud_id]/resolver
  if (solicitud_id && action === 'resolver') {
    if (req.method !== 'POST') return error(res, 'Método no permitido', 405)
    const { decision, monto, plazo, motivo } = req.body || {}
    if (!decision) return error(res, 'decision requerida', 400)

    const { error: rpcErr } = await sb.rpc('resolver_solicitud', {
      p_solicitud_id: solicitud_id,
      p_decision: decision.toUpperCase(),
      p_monto_aprobado: monto ?? null,
      p_plazo: plazo ?? null,
      p_motivo: motivo ?? null,
    })
    if (rpcErr) return error(res, rpcErr.message, 500)
    return res.status(200).json({ ok: true })
  }

  // GET/POST /api/solicitudes/[solicitud_id]/notas
  if (solicitud_id && action === 'notas') {
    if (req.method === 'GET') {
      const { data, error: dbErr } = await sb
        .from('solicitudes_notas_internas')
        .select('*')
        .eq('solicitud_id', solicitud_id)
        .order('created_at', { ascending: false })
      if (dbErr) return error(res, dbErr.message, 500)
      return res.status(200).json(
        (data || []).map((n) => ({ contenido: n.nota, created_at: n.created_at }))
      )
    }

    if (req.method === 'POST') {
      const { contenido } = req.body || {}
      if (!contenido) return error(res, 'contenido requerido', 400)

      const { error: insErr } = await sb.from('solicitudes_notas_internas').insert({
        solicitud_id,
        asesor_id: user.id,
        nota: contenido,
      })
      if (insErr) return error(res, insErr.message, 500)
      return res.status(200).json({ ok: true })
    }

    return error(res, 'Método no permitido', 405)
  }

  // GET /api/solicitudes
  if (req.method === 'GET') {
    const { data, error: dbErr } = await sb
      .from('solicitudes')
      .select('*')
      .eq('asesor_id', user.id)
      .order('created_at', { ascending: false })
    if (dbErr) return error(res, dbErr.message, 500)
    return res.status(200).json((data || []).map(mapSolicitud))
  }

  // POST /api/solicitudes
  if (req.method === 'POST') {
    const body = req.body || {}
    const nombre = `${body.nombres || ''} ${body.apellidos || ''}`.trim()

    const { data, error: rpcErr } = await sb.rpc('crear_solicitud_desde_asesor', {
      p_documento: (body.numero_documento || '').trim(),
      p_nombres: nombre,
      p_monto: body.monto_solicitado,
      p_plazo: body.plazo_meses || 12,
      p_telefono: body.telefono || null,
      p_tipo_negocio: body.tipo_negocio || null,
      p_nombre_negocio: body.nombre_negocio || null,
      p_destino: body.destino_credito || null,
      p_tea: body.tea_referencial ?? 43.92,
    })

    if (rpcErr) return error(res, rpcErr.message, 500)
    const d = data || {}
    return res.status(200).json({
      id: d.id,
      numero_expediente: d.expediente_numero || d.numero_expediente || d.id,
      expediente_numero: d.expediente_numero || d.id,
      estado: (d.estado || 'enviado').toLowerCase(),
      cuota_mensual: d.cuota_mensual || null,
    })
  }

  return error(res, 'Método no permitido', 405)
}

File: api/routers/solicitudes.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from deps import get_current_user, sb_user
from services import map_cliente, map_solicitud

router = APIRouter(tags=["solicitudes"])


class SolicitudBody(BaseModel):
    numero_documento: str
    nombres: str
    apellidos: str = ""
    telefono: str | None = None
    tipo_negocio: str | None = None
    nombre_negocio: str | None = None
    ingresos_estimados: float | None = None
    monto_solicitado: float
    plazo_meses: int = 12
    destino_credito: str | None = None
    tea_referencial: float = 43.92


class ResolverBody(BaseModel):
    decision: str
    monto: float | None = None
    plazo: int | None = None
    motivo: str | None = None


class NotaBody(BaseModel):
    contenido: str


@router.get("/solicitudes")
def listar(user=Depends(get_current_user), sb=Depends(sb_user)):
    res = (
        sb.table("solicitudes")
        .select("*")
        .eq("asesor_id", user["id"])
        .order("created_at", desc=True)
        .execute()
    )
    return [map_solicitud(r) for r in (res.data or [])]


@router.post("/solicitudes")
def crear(body: SolicitudBody, user=Depends(get_current_user), sb=Depends(sb_user)):
    nombre = f"{body.nombres} {body.apellidos}".strip()
    res = sb.rpc(
        "crear_solicitud_desde_asesor",
        {
            "p_documento": body.numero_documento.strip(),
            "p_nombres": nombre,
            "p_monto": body.monto_solicitado,
            "p_plazo": body.plazo_meses,
            "p_telefono": body.telefono,
            "p_tipo_negocio": body.tipo_negocio,
            "p_nombre_negocio": body.nombre_negocio,
            "p_destino": body.destino_credito,
            "p_tea": body.tea_referencial,
        },
    ).execute()
    data = res.data or {}
    return {
        "numero_expediente": data.get("expediente_numero") or data.get("id"),
        "estado": "enviado",
        "id": data.get("id"),
    }


@router.post("/solicitudes/{solicitud_id}/resolver")
def resolver(
    solicitud_id: str,
    body: ResolverBody,
    user=Depends(get_current_user),
    sb=Depends(sb_user),
):
    sb.rpc(
        "resolver_solicitud",
        {
            "p_solicitud_id": solicitud_id,
            "p_decision": body.decision.upper(),
            "p_monto_aprobado": body.monto,
            "p_plazo": body.plazo,
            "p_motivo": body.motivo,
        },
    ).execute()
    return {"ok": True}


@router.get("/solicitudes/{solicitud_id}/notas")
def listar_notas(solicitud_id: str, user=Depends(get_current_user), sb=Depends(sb_user)):
    res = (
        sb.table("solicitudes_notas_internas")
        .select("*")
        .eq("solicitud_id", solicitud_id)
        .order("created_at", desc=True)
        .execute()
    )
    return [{"contenido": n["nota"], "created_at": n["created_at"]} for n in (res.data or [])]


@router.post("/solicitudes/{solicitud_id}/notas")
def agregar_nota(
    solicitud_id: str,
    body: NotaBody,
    user=Depends(get_current_user),
    sb=Depends(sb_user),
):
    sb.table("solicitudes_notas_internas").insert(
        {
            "solicitud_id": solicitud_id,
            "asesor_id": user["id"],
            "nota": body.contenido,
        }
    ).execute()
    return {"ok": True}


@router.get("/clientes/{cliente_id}/ficha")
def ficha_cliente(cliente_id: str, user=Depends(get_current_user), sb=Depends(sb_user)):
    cliente = (
        sb.table("clientes").select("*").eq("id", cliente_id).limit(1).execute()
    )
    if not cliente.data:
        raise HTTPException(status_code=404, detail="Cliente no encontrado")
    c = cliente.data[0]
    creditos = sb.table("creditos").select("*").eq("cliente_id", cliente_id).execute()
    solicitudes = (
        sb.table("solicitudes")
        .select("*")
        .eq("cliente_id", cliente_id)
        .order("created_at", desc=True)
        .limit(5)
        .execute()
    )
    return {
        "cliente": map_cliente(c),
        "creditos": creditos.data or [],
        "solicitudes": [map_solicitud(s) for s in (solicitudes.data or [])],
        "oferta": {
            "monto_preaprobado": c.get("monto_preaprobado"),
            "plazo": c.get("plazo_preaprobado"),
            "tasa": c.get("tasa_preaprobada"),
            "segmento": c.get("segmento"),
        },
    }
